File: scripts/ai-workflow/core/phase_dependencies.py
"""フェーズ依存関係管理モジュール

各フェーズの依存関係を定義し、実行前に依存関係をチェックする機能を提供します。

主要機能:
- PHASE_DEPENDENCIES: フェーズ依存関係の定義
- PHASE_PRESETS: プリセット実行モードの定義
- validate_phase_dependencies(): 依存関係検証
- detect_circular_dependencies(): 循環参照検出
- validate_external_document(): 外部ドキュメント検証
"""
from pathlib import Path
from typing import Dict, List, Any, Optional


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# フェーズ依存関係定義
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

PHASE_DEPENDENCIES: Dict[str, List[str]] = {
    'planning': [],  # 依存なし
    'requirements': ['planning'],
    'design': ['requirements'],
    'test_scenario': ['requirements', 'design'],
    'implementation': ['requirements', 'design', 'test_scenario'],
    'test_implementation': ['implementation'],
    'testing': ['test_implementation'],
    'documentation': ['implementation'],
    'report': ['requirements', 'design', 'implementation', 'testing', 'documentation'],
    'evaluation': ['report']
}


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# プリセット定義
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

PHASE_PRESETS: Dict[str, List[str]] = {
    'requirements-only': ['requirements'],
    'design-phase': ['requirements', 'design'],
    'implementation-phase': ['requirements', 'design', 'test_scenario', 'implementation'],
    'full-workflow': [
        'planning', 'requirements', 'design', 'test_scenario',
        'implementation', 'test_implementation', 'testing',
        'documentation', 'report', 'evaluation'
    ]
}


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# 依存関係検証関数
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

def validate_phase_dependencies(
    phase_name: str,
    metadata_manager,
    skip_check: bool = False,
    ignore_violations: bool = False
) -> Dict[str, Any]:
    """
    フェーズ実行前に依存関係をチェック

    Args:
        phase_name: フェーズ名（例: 'implementation'）
        metadata_manager: MetadataManagerインスタンス
        skip_check: 依存関係チェックをスキップするか（--skip-dependency-check）
        ignore_violations: 依存関係違反を警告のみで許可するか（--ignore-dependencies）

    Returns:
        Dict[str, Any]: 検証結果
            - valid: bool - 依存関係が満たされているか
            - error: Optional[str] - エラーメッセージ（valid=False の場合）
            - warning: Optional[str] - 警告メッセージ（ignored=True の場合）
            - ignored: bool - 依存関係違反が無視されたか
            - missing_phases: List[str] - 未完了の依存フェーズ一覧

    Raises:
        ValueError: phase_name が不正な場合

    Example:
        >>> result = validate_phase_dependencies('implementation', metadata_manager)
        >>> if not result['valid']:
        ...     print(result['error'])
        Phase 'requirements' must be completed before 'implementation'
    """
    # フェーズ名のバリデーション
    if phase_name not in PHASE_DEPENDENCIES:
        raise ValueError(f"Invalid phase name: {phase_name}")

    # skip_check=True の場合は即座にリターン
    if skip_check:
        return {'valid': True}

    # 依存フェーズリストを取得
    required_phases = PHASE_DEPENDENCIES.get(phase_name, [])

    # 依存関係がない場合（planningフェーズ）
    if not required_phases:
        return {'valid': True}

    # 全フェーズのステータスを取得
    phases_status = metadata_manager.get_all_phases_status()

    # 未完了の依存フェーズをチェック
    missing_phases = []
    for required_phase in required_phases:
        status = phases_status.get(required_phase)
        if status != 'completed':
            missing_phases.append(required_phase)
            # 早期リターン最適化（ignore_violationsがFalseの場合）
            if not ignore_violations:
                return {
                    'valid': False,
                    'error': f"Phase '{required_phase}' must be completed before '{phase_name}'",
                    'missing_phases': [required_phase]
                }

    # すべてチェック完了
    if missing_phases and ignore_violations:
        return {
            'valid': False,
            'ignored': True,
            'warning': f"Dependency violations ignored: {', '.join(missing_phases)}",
            'missing_phases': missing_phases
        }

    return {'valid': True}


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# 循環参照検出関数
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

def detect_circular_dependencies() -> List[List[str]]:
    """
    PHASE_DEPENDENCIES 内の循環参照を検出

    Returns:
        List[List[str]]: 循環参照のリスト（例: [['A', 'B', 'A']]）
                         循環参照が存在しない場合は空リスト

    Algorithm:
        DFS (Depth-First Search) を使用して循環を検出

    Example:
        >>> cycles = detect_circular_dependencies()
        >>> if cycles:
        ...     print(f"Circular dependencies detected: {cycles}")
    """
    def dfs(node: str, visited: set, rec_stack: list, path: list) -> List[List[str]]:
        """DFSで循環参照を検出"""
        visited.add(node)
        rec_stack.append(node)
        path.append(node)
        cycles = []

        dependencies = PHASE_DEPENDENCIES.get(node, [])
        for dep in dependencies:
            if dep not in visited:
                # 未訪問ノードを訪問
                cycles.extend(dfs(dep, visited, rec_stack, path))
            elif dep in rec_stack:
                # 循環検出
                cycle_start_index = rec_stack.index(dep)
                cycle = rec_stack[cycle_start_index:] + [dep]
                cycles.append(cycle)

        # バックトラック
        rec_stack.pop()
        path.pop()
        return cycles

    visited = set()
    all_cycles = []

    # すべてのノードを訪問
    for phase in PHASE_DEPENDENCIES:
        if phase not in visited:
            all_cycles.extend(dfs(phase, visited, [], []))

    return all_cycles


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# 外部ドキュメント検証関数
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

def validate_external_document(file_path: str, repo_root: Optional[Path] = None) -> Dict[str, Any]:
    """
    外部ドキュメントファイルのバリデーション

    Args:
        file_path: ファイルパス（相対パスまたは絶対パス）
        repo_root: リポジトリルートパス（省略可、指定しない場合は相対パスベース）

    Returns:
        Dict[str, Any]: バリデーション結果
            - valid: bool
            - error: Optional[str]
            - absolute_path: str - 絶対パス

    Validation rules:
        - ファイルが存在すること
        - 読み込み可能な形式（.md, .txt）
        - ファイルサイズが10MB以下
        - リポジトリ内のファイル（相対パスで指定）

    Example:
        >>> result = validate_external_document('path/to/requirements.md')
        >>> if not result['valid']:
        ...     print(result['error'])
    """
    try:
        # Pathオブジェクトに変換
        file = Path(file_path)

        # ファイルの存在確認
        if not file.exists():
            return {
                'valid': False,
                'error': f'File not found: {file_path}'
            }

        # 絶対パスに変換
        abs_path = file.resolve()

        # ファイル拡張子のチェック（.md, .txt のみ許可）
        if abs_path.suffix not in ['.md', '.txt']:
            return {
                'valid': False,
                'error': f'Invalid file format: {abs_path.suffix}. Only .md and .txt are allowed'
            }

        # ファイルサイズのチェック（10MB以下）
        file_size_mb = abs_path.stat().st_size / (1024 * 1024)
        if file_size_mb > 10.0:
            return {
                'valid': False,
                'error': f'File size exceeds 10MB limit (actual: {file_size_mb:.1f}MB)'
            }

        # リポジトリ内のファイルかチェック（セキュリティ）
        if repo_root:
            repo_root_resolved = repo_root.resolve()
            try:
                abs_path.relative_to(repo_root_resolved)
            except ValueError:
                return {
                    'valid': False,
                    'error': 'File must be within the repository'
                }

        return {
            'valid': True,
            'absolute_path': str(abs_path)
        }

    except PermissionError:
        return {
            'valid': False,
            'error': f'Permission denied: {file_path}'
        }
    except Exception as e:
        return {
            'valid': False,
            'error': f'Unexpected error: {str(e)}'
        }

File: scripts/ai-workflow-v2/src/core/phase-dependencies.ts
import fs from 'fs-extra';
import { resolve as resolvePath } from 'node:path';
import { MetadataManager } from './metadata-manager.js';
import { PhaseName, PhaseStatus } from '../types.js';

export const PHASE_DEPENDENCIES: Record<PhaseName, PhaseName[]> = {
  planning: [],
  requirements: ['planning'],
  design: ['requirements'],
  test_scenario: ['requirements', 'design'],
  implementation: ['requirements', 'design', 'test_scenario'],
  test_implementation: ['implementation'],
  testing: ['test_implementation'],
  documentation: ['implementation'],
  report: ['requirements', 'design', 'implementation', 'testing', 'documentation'],
  evaluation: ['report'],
};

// 新規プリセット定義（Issue #396）
export const PHASE_PRESETS: Record<string, PhaseName[]> = {
  // === レビュー駆動パターン ===
  'review-requirements': ['planning', 'requirements'],
  'review-design': ['planning', 'requirements', 'design'],
  'review-test-scenario': ['planning', 'requirements', 'design', 'test_scenario'],

  // === 実装中心パターン ===
  'quick-fix': ['implementation', 'documentation', 'report'],
  'implementation': ['implementation', 'test_implementation', 'testing', 'documentation', 'report'],

  // === テスト中心パターン ===
  'testing': ['test_implementation', 'testing'],

  // === ドキュメント・レポートパターン ===
  'finalize': ['documentation', 'report', 'evaluation'],
};

// 後方互換性のための非推奨プリセット（6ヶ月後に削除予定）
export const DEPRECATED_PRESETS: Record<string, string> = {
  'requirements-only': 'review-requirements',
  'design-phase': 'review-design',
  'implementation-phase': 'implementation',
  'full-workflow': '--phase all',
};

// プリセット説明マップ
export const PRESET_DESCRIPTIONS: Record<string, string> = {
  'review-requirements': 'Planning + Requirements (要件定義レビュー用)',
  'review-design': 'Planning + Requirements + Design (設計レビュー用)',
  'review-test-scenario': 'Planning + Requirements + Design + TestScenario (テストシナリオレビュー用)',
  'quick-fix': 'Implementation + Documentation + Report (軽微な修正用)',
  'implementation': 'Implementation + TestImplementation + Testing + Documentation + Report (通常の実装フロー)',
  'testing': 'TestImplementation + Testing (テスト追加用)',
  'finalize': 'Documentation + Report + Evaluation (最終化用)',
};

export interface DependencyValidationOptions {
  skipCheck?: boolean;
  ignoreViolations?: boolean;
  checkFileExistence?: boolean; // ファイル存在チェック（Issue #396）
  presetPhases?: PhaseName[]; // プリセット実行時のフェーズリスト（Issue #396）
}

export interface DependencyValidationResult {
  valid: boolean;
  error?: string;
  warning?: string;
  ignored?: boolean;
  missing_phases?: PhaseName[];
  missing_files?: Array<{ phase: PhaseName; file: string }>; // ファイル不在情報（Issue #396）
}

export const validatePhaseDependencies = (
  phaseName: PhaseName,
  metadataManager: MetadataManager,
  options: DependencyValidationOptions = {},
): DependencyValidationResult => {
  const { skipCheck = false, ignoreViolations = false, checkFileExistence = false, presetPhases } = options;

  if (!(phaseName in PHASE_DEPENDENCIES)) {
    throw new Error(`Invalid phase name: ${phaseName}`);
  }

  if (skipCheck) {
    return { valid: true, missing_files: [] };
  }

  const required = PHASE_DEPENDENCIES[phaseName] ?? [];
  if (required.length === 0) {
    return { valid: true, missing_files: [] };
  }

  // プリセット実行時は、プリセットに含まれるフェーズのみを依存関係としてチェック
  const filteredRequired = presetPhases
    ? required.filter(dep => presetPhases.includes(dep))
    : required;

  if (filteredRequired.length === 0) {
    return { valid: true, missing_files: [] };
  }

  const missing: PhaseName[] = [];
  const missingFiles: Array<{ phase: PhaseName; file: string }> = [];

  // 各依存Phaseをチェック
  for (const depPhase of filteredRequired) {
    let status: PhaseStatus;
    try {
      status = metadataManager.getPhaseStatus(depPhase);
    } catch {
      missing.push(depPhase);
      continue;
    }

    if (status !== 'completed') {
      missing.push(depPhase);
      continue;
    }

    // ファイル存在チェック（オプション）
    if (checkFileExistence) {
      const expectedFile = getPhaseOutputFilePath(depPhase, metadataManager);
      if (expectedFile && !fs.existsSync(expectedFile)) {
        missingFiles.push({ phase: depPhase, file: expectedFile });
      }
    }
  }

  // チェック結果の判定
  const hasViolations = missing.length > 0 || missingFiles.length > 0;

  if (!hasViolations) {
    return { valid: true, missing_files: [] };
  }

  // ignoreViolationsの場合は警告のみ
  if (ignoreViolations) {
    const warningMessage = buildWarningMessage(phaseName, missing, missingFiles);
    return {
      valid: true,
      warning: warningMessage,
      ignored: true,
      missing_phases: missing,
      missing_files: missingFiles,
    };
  }

  // エラーメッセージを構築
  const errorMessage = buildErrorMessage(phaseName, missing, missingFiles);
  return {
    valid: false,
    error: errorMessage,
    missing_phases: missing,
    missing_files: missingFiles,
  };
};

/**
 * エラーメッセージを構築
 */
function buildErrorMessage(
  phaseName: PhaseName,
  missingDependencies: PhaseName[],
  missingFiles: Array<{ phase: PhaseName; file: string }>,
): string {
  let message = `[ERROR] Phase "${phaseName}" requires the following phases to be completed:\n`;

  // 未完了Phaseのリスト
  for (const dep of missingDependencies) {
    message += `  ✗ ${dep} - NOT COMPLETED\n`;
  }

  // ファイル不在のリスト
  for (const { phase, file } of missingFiles) {
    message += `  ✗ ${phase} - ${file} NOT FOUND\n`;
  }

  message += `\nOptions:\n`;
  message += `  1. Complete the missing phases first\n`;
  message += `  2. Use --phase all to run all phases\n`;
  message += `  3. Use --ignore-dependencies to proceed anyway (not recommended)\n`;

  return message;
}

/**
 * 警告メッセージを構築
 */
function buildWarningMessage(
  phaseName: PhaseName,
  missingDependencies: PhaseName[],
  missingFiles: Array<{ phase: PhaseName; file: string }>,
): string {
  let message = `[WARNING] Phase "${phaseName}" has unmet dependencies, but proceeding anyway...\n`;

  // 未完了Phaseのリスト
  for (const dep of missingDependencies) {
    message += `  ⚠ ${dep} - NOT COMPLETED\n`;
  }

  // ファイル不在のリスト
  for (const { phase, file } of missingFiles) {
    message += `  ⚠ ${phase} - ${file} NOT FOUND\n`;
  }

  return message;
}

/**
 * Phase出力ファイルのパスを取得
 */
function getPhaseOutputFilePath(phaseName: PhaseName, metadataManager: MetadataManager): string | null {
  const phaseNumberMap: Record<PhaseName, string> = {
    'planning': '00_planning',
    'requirements': '01_requirements',
    'design': '02_design',
    'test_scenario': '03_test_scenario',
    'implementation': '04_implementation',
    'test_implementation': '05_test_implementation',
    'testing': '06_testing',
    'documentation': '07_documentation',
    'report': '08_report',
    'evaluation': '09_evaluation',
  };

  const fileNameMap: Record<PhaseName, string> = {
    'planning': 'planning.md',
    'requirements': 'requirements.md',
    'design': 'design.md',
    'test_scenario': 'test-scenario.md',
    'implementation': 'implementation.md',
    'test_implementation': 'test-implementation.md',
    'testing': 'test-result.md',
    'documentation': 'documentation-update-log.md',
    'report': 'report.md',
    'evaluation': 'evaluation.md',
  };

  const phaseDir = phaseNumberMap[phaseName];
  const fileName = fileNameMap[phaseName];

  if (!phaseDir || !fileName) {
    return null;
  }

  return resolvePath(
    metadataManager.workflowDir,
    phaseDir,
    'output',
    fileName,
  );
}

export const detectCircularDependencies = (): PhaseName[][] => {
  const visited = new Set<PhaseName>();
  const stack: PhaseName[] = [];
  const cycles: PhaseName[][] = [];

  const dfs = (phase: PhaseName) => {
    visited.add(phase);
    stack.push(phase);

    for (const dependency of PHASE_DEPENDENCIES[phase] ?? []) {
      if (!visited.has(dependency)) {
        dfs(dependency);
      } else if (stack.includes(dependency)) {
        const cycleStart = stack.indexOf(dependency);
        cycles.push([...stack.slice(cycleStart), dependency]);
      }
    }

    stack.pop();
  };

  (Object.keys(PHASE_DEPENDENCIES) as PhaseName[]).forEach((phase) => {
    if (!visited.has(phase)) {
      dfs(phase);
    }
  });

  return cycles;
};

export interface ExternalDocumentValidation {
  valid: boolean;
  absolute_path?: string;
  error?: string;
}

export const validateExternalDocument = (
  filePath: string,
  repoRoot?: string,
): ExternalDocumentValidation => {
  try {
    const absolutePath = resolvePath(filePath);

    if (!fs.existsSync(absolutePath)) {
      return { valid: false, error: `File not found: ${filePath}` };
    }

    const extension = absolutePath.split('.').pop()?.toLowerCase();
    if (!extension || !['md', 'txt'].includes(extension)) {
      return {
        valid: false,
        error: `Invalid file format: .${extension ?? 'unknown'}. Only .md and .txt are allowed`,
      };
    }

    const stats = fs.statSync(absolutePath);
    if (stats.size > 10 * 1024 * 1024) {
      return {
        valid: false,
        error: `File size exceeds 10MB limit (actual: ${(stats.size / (1024 * 1024)).toFixed(1)}MB)`,
      };
    }

    if (repoRoot) {
      const normalizedRoot = resolvePath(repoRoot);
      if (!resolvePath(absolutePath).startsWith(normalizedRoot)) {
        return {
          valid: false,
          error: 'File must be within the repository',
        };
      }
    }

    return {
      valid: true,
      absolute_path: absolutePath,
    };
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'EACCES') {
      return { valid: false, error: `Permission denied: ${filePath}` };
    }
    return { valid: false, error: `Unexpected error: ${(error as Error).message}` };
  }
};
